backends/nxp/neutron_partitioner.py

```diff
@@ -0,0 +1,327 @@
+# Copyright (c) 2024-2025 NXP
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Partitioner for the NXP Neutron NPU
+
+import logging
+import operator
+from dataclasses import dataclass
+from typing import Dict, final, List, Mapping
+
+import torch
+
+from executorch.backends.nxp.backend.edge_program_converter import (
+    EdgeProgramToIRConverter,
+)
+from executorch.backends.nxp.backend.ir.converter.node_converter import Target
+from torch.export.exported_program import ExportedProgram
+from torch.fx.passes.infra.partitioner import CapabilityBasedPartitioner
+from torch.fx.passes.operator_support import OperatorSupportBase
+from torch.nn import Parameter
+from executorch.backends.nxp.backend.ir.converter.node_converters.ops_converters import *
+from executorch.backends.nxp.nxp_backend import NeutronBackend
+from executorch.exir.backend.compile_spec_schema import CompileSpec
+from executorch.exir.backend.partitioner import (
+    DelegationSpec,
+    Partitioner,
+    PartitionResult,
+)
+from executorch.exir.backend.utils import tag_constant_data
+from executorch.exir.dialects._ops import ops as exir_ops
+
+
+class QDQClusterRecognizer:
+    """
+    Implementation of the Quantize - Dequantize clustering.
+    The quantization is captured in the ExecuTorch program using the QDQ (Quantize - DeQuantize) representation. Here
+    the inputs to a node comes from some dequantize nodes and outputs goes to some quantize nodes.
+    The QDQClusterRecognizer identifies operator performing the quantized arithmetic represented in QDQ form, and the
+    corresponding QDQ cluster. The QDQ cluster consists of the:
+    - dequantize nodes producing the inputs to the compute node
+    - compute node (e.g. conv)
+    - auxiliary nodes, like getitem, view_copy, ... which does not perform a core computation
+    - quantize nodes processing the output of the compute node.
+    """
+
+    @dataclass
+    class QDQCluster:
+        """
+        Dataclass to hold the QDQ cluster instance. For the purpose of Partitioner we hold the list of operators,
+        in the QDQ cluster (`ops`) and the compute node what the QDQ cluster is built around.
+        The compute node is what is represented in the Neutron IR. the rest of nodes are helpers for data transformation,
+        and defines the quantization parameters. This gives the partitioner the ability to:
+            - identify if the node is part of a QDQ cluster
+            - reference the compute node in the QDQ cluster
+        """
+
+        compute_node: torch.fx.Node
+        ops: List[torch.fx.Node]
+
+    QUANTIZE_OPERATORS = [
+        exir_ops.edge.quantized_decomposed.quantize_per_channel.default,
+        exir_ops.edge.quantized_decomposed.quantize_per_tensor.default,
+        exir_ops.edge.quantized_decomposed.quantize_per_tensor.tensor,
```

Comment:
This is for dynamic (at runtime) quantization of the activation for conv, linear etc. Do we support that?

Reply:
I guess the best way to check for the dynamic quant is presence of the `choose_qparam` node to dynamically populate scale and zp. I am not sure if we are being consistent with `.tensor` variant TBH.